[Heyho] 스프링 카페 1단계 - 회원 가입 및 목록 기능

dongkyu-park opened the pull request:
안녕하세요. heyho입니다.
스프링 카페 1단계 기능 구현을 완료하여 PR요청 드립니다.
감사합니다.

dongkyu-park committed:
feat: 메인화면 지정 및 로그인, 회원가입, 회원목록 페이지 이동 기능

로그인 버튼 클릭 시 로그인 페이지로 이동
회원가입 버튼 클릭 시 회원가입 페이지로 이동
사람 모양 버튼 클릭 시 회원목록 페이지로 이동

dongkyu-park committed:
feat: html 중복코드 제거 및 profile 페이지 이동 기능

Home navi를 클릭하고 Profile 버튼 클릭 시 유저 상세 정보 페이지로 이동한다.

dongkyu-park committed:
feat: 회원가입 기능 추가

dongkyu-park committed:
feat: 회원 목록 조회 기능 추가

dongkyu-park committed:
feat: 회원 프로필 상세 조회 기능 추가

dongkyu-park committed:
fix: 네비게이션 바 profile 페이지 이동 링크 제거

추후 로그인 기능 구현 후 추가해야함.

dongkyu-park committed:
feat: 이미 가입 된 아이디로 회원가입을 시도 할 경우 가입이 불가능하도록 변경

dongkyu-park committed:
refactor: User의 setter메서드를 없애고 builder패턴으로 객체를 생성하도록 변경

dongkyu-park committed:
refactor: html파일 라인 끝에 개행문자 추가

dongkyu-park committed:
refactor: 전달받은 userId를 user객체로 전달하여 객체가 직접 자신의 userId와 일치하는지 확인하게끔 변경

dongkyu-park committed:
test: User.class -> 동일한 아이디를 확인하는 메서드의 테스트 코드 추가

wheejuni commented on src/main/java/com/kakao/cafe/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class AppConfig {
```
보통 이렇게 컨피규레이션 클래스를 작성하는 경우에는 계층별로, 혹은 기능별로 클래스를 나눠서 선언하는 편입니다.
`UserRepository` 에 대한 configuration이라면 그 기능에 맞는 클래스 이름을 가지면 좋겠네요!

dongkyu-park replied:
단순히 어플리케이션의 규모가 커져 여러 bean 설정을 해주어야 할 경우 `@Component` 설정을 한 곳에 모으는 것이 편하겠다는 것에만 초점을 맞춰 생각하고 진행 했었습니다. 하나의 자바 클래스파일에 모든 비즈니스 로직이 들어가는 것과 같은 구조였네요. 감사합니다!!

wheejuni replied:
네 감사합니다. 참고로 애노테이션을 표기하실 때는 백틱 (` `) 으로 감싸시는 것이 다른 사람이 태그되는 참사를 피하는 길입니다 ㅎㅎ

wheejuni replied:
헐... ` 입니다.

dongkyu-park replied:
앗 이해했습니다 감사합니다!

wheejuni commented on src/main/java/com/kakao/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+
+@Controller
+public class UserController {
+    UserService userService;
```
접근제어자가 누락되어 있군요

dongkyu-park replied:
앗.. 감사합니다

wheejuni commented on src/main/java/com/kakao/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+public class UserController {
+    UserService userService;
+
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
```
오 생성자 주입 좋습니다 💯

wheejuni commented on src/main/java/com/kakao/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public User save(User user) {
+        store.add(user);
+        return store.get(store.size() - 1);
```
이건 어떨까요?

```suggestion
        return user;

dongkyu-park replied:
```java
public User save(User user) {
    store.add(user);
    return user;
}
```

list에 저장되는 Object가 깊은복사로 들어간다고 생각하여 store에 저장되는 user객체와 다른 user를 리턴하는 것이 석연치 않다고 생각하여 해당 방식으로 구현을 했었습니다. 

방금 테스트를 해보니 Object는 CallByReference개념으로.. 얕은복사로 들어간다는 것을 알게 되었습니다. 하하..😂 잘못 된 생각을 바로 잡을 수 있었어요. 감사합니다..!!

dongkyu-park replied:
질문 한 가지 드리고 싶습니다. 강의 예제 따라치기 하느라 생각없이 구현한 부분이였는데.. save후 왜 객체를 반환하는가? 에 대한 주제가 모각코방에서 나왔었는데, 명확한 이유를 찾지 못 하겠습니다. 지금 단계에선 필요성이 없어 보이고, jpa와 연관된 것일까?하는 추측만 하고 있는데, 어떤 키워드로 찾아 학습하면 좋을 지 조언을 얻을 수 있을까요?

wheejuni replied:
사실 이건 학습이라기보단 공개된 API를 설계하는 관점이나 시각? 생각? 에 관한 문제일 것 같아요.
`내 코드를 쓰는 사람이 내가 아니라면` 이라는 상상에서 출발할 것 같습니다.

wheejuni commented on src/main/java/com/kakao/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,35 @@
+        return Collections.unmodifiableList(store);
+    }
+
+    @Override
+    public void clear() {
+        store.clear();
+    }
```
오 엄청나게 위험한 기능인데 `public` 으로 오픈해도 괜찮나요?

dongkyu-park replied:
실제로 서비스하는 어플리케이션에서 데이터 저장을 jvm메모리에 올린 채 서비스 될 일은 없다 생각하였고, 메모리 저장소는 개발자가 어플리케이션을 만드는 개발 과정에서 테스트 용도로만 사용될 것이라 생각했기에 깊이 생각하지 않고 초기화 시키는 기능을 넣었습니다. 

사실 테스트 코드를 작성하는 과정에서 추가하게 된 메서드인데 MemoryRepository에서 User가 저장 될 store를 static으로 선언하여 관리하고 있는 상황인데 `public` 오픈을 없애고 다른 방식으로 개선하고자 한다면 어떠한 방법을 사용할 수 있을까요?

wheejuni replied:
테스트 코드를 작성하는 과정에서 메서드가 추가되면 안됩니다.

wheejuni commented on src/main/java/com/kakao/cafe/repository/UserRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<User> findAll();
+
+    default void clear() {};
```
흠 이거 의도를 조금 더 알려주실 수 있을까요?

dongkyu-park replied:
```java
class UserServiceTest {

    private UserService userService;
    private UserRepository userRepository;

    @BeforeEach
    public void beforeEach() {
        userRepository = new MemoryUserRepository();
        userService = new UserService(userRepository);
    }

    @AfterEach
    public void afterEach() {
        userRepository.clear();
    }
```

앗.. 이건 다른 저장소를 구현체로 만들게 될 경우 데이터를 날려버리는... 기능이 필요하다면 구현하고, 아니면 사용하지 않고의 느낌으로 추가한 부분이였는데 다시 생각해보니 필요성도 없고 무의미한 것 같습니다..😱 

위에서 드렸던 질문과 이어지는 내용인 것 같은데 인터페이스에서 clear()를 제거하고 나면, UserServiceTest 코드에서 clear()를 사용하고자 한다면 `private MemoryUserRepository userRepository;` 로 필드변수의 타입을 구체적으로 명시 해줘야 하는 상황인데, 테스트 코드이기 때문에 괜찮을까요? 아니면 구조가 잘못 된 걸까요?

wheejuni replied:
테스트 코드를 위해 메서드를 추가하면 안되고 스펙은 더더욱 추가하면 안됩니다.
안전한 테스트를 위한 다른 방법을 고민해 보심이 어떠한지요?

wheejuni reviewed (changes requested):
수고 많으셨습니다 💯 
몇가지 질문과 코멘트 남겨 보았어요. 확인 부탁드려요.

dongkyu-park committed:
refactor: AppConfig.class -> RepositoryConfig.class 클래스 파일명 변경

dongkyu-park committed:
refactor: 누락된 접근제어자 추가

dongkyu-park committed:
refactor: 메모리 저장소의 저장메서드 반환값 변경

객체를 list에 저장 후 list의 마지막 값으로 반환하던 것을 -> 저장 후 사용한 객체 자체를 반환 하도록 변경

dongkyu-park committed:
refactor: UserRepository 인터페이스의 clear() 추상메서드 삭제

dongkyu-park committed:
refactor: 필드 변수 final 키워드 추가

필드에서 생성을 해주지 않았거나, 생성자에 실수로 값을 넣어주지 않거나 실수 할 때 컴파일 시점에서 오류를 발생시켜 오동작을 막을 수 있다.
다른 메서드 로직에서 값을 임의로 변경시키지 않도록 막을 수 있다.

wheejuni commented:
1단계에서 너무 오래 있는 것 같아 우선 머지합니다~ 코멘트 남겨드린 부분은 고민해주세요. 코멘트가 늦었군요 ㅠ 
@dongkyu-park

dongkyu-park commented:
감사합니다😊
